utils/ftpcount.1.in
.TH ftpcount 1 "March 2003"
.\" Process with
.\" groff -man -Tascii ftpcount.1 
.\"
.SH NAME
ftpcount \- show current number of connections for each proftpd server
configuration
.SH SYNOPSIS
.B ftpcount
.SH DESCRIPTION
The
.BI ftpcount
command shows the current number of connections per server and
virtualhost/anonymous configuration defined in the proftpd.conf file.
Connections spawned by inetd are counted separately from those created
by a master proftpd
.I standalone
server.
.SH OPTIONS
.TP 12
.B \-h,\--help
Display a short usage description, including all available options.
.TP
.BI \-f,\--file " scoreboard\-file"
Specify the full path to proftpd's run\-time scoreboard file (configured
via the \fBScoreboardFile\fP directive in
.I proftpd.conf
).  If proftpd's
default directory has been changed via this directive, \fBftpcount\fP
must either be recompiled or this option must be used in order to find
proftpd's scoreboard.
.TP
.BI \-S,\--server " server\-name"
Specify the ServerName for a specific virtual host.  If used,
.I ftpcount
will only show the session count for the given virtual host.
.SH FILES
.PD 0
.B @BINDIR@/ftpcount
.br
.B @LOCALSTATEDIR@/proftpd/proftpd.scoreboard
.PD
.SH AUTHORS
.PP
ProFTPD is written and maintained by a number of people, full credits
can be found on
.BR http://www.proftpd.org/credits.html
.PD
.SH SEE ALSO
.BR inetd(8), ftp(1), proftpd(8), ftpwho(1), ftptop(1), ftpshut(8)
.PP
Full documentation on ProFTPD, including configuration and FAQs, is available at
.BR http://www.proftpd.org/
.PP 
For help/support, try the ProFTPD mailing lists, detailed on
.BR http://www.proftpd.org/lists.html
.PP
Report bugs at
.BR http://bugs.proftpd.org/


NEWS
$Id: NEWS,v 1.1009 2009-03-12 17:11:15 castaglia Exp $

-----------------------------------------------------------------------------
  More details on the bugs listed below can be found by using the bug number
  indicated in the following URL:

    http://bugs.proftpd.org/show_bug.cgi?id=N

  where `N' is the bug number.
-----------------------------------------------------------------------------

CVS
--------------------------------
- Bug 1801 - Faulty SQLConnectInfo PERCALL policy - connections not closed
  after each call.
- Bug 2894 - Deprecate the AnonymousGroup directive.
- Bug 3123 - Use PQescapeStringConn() rather than PQescapeString().
- Bug 3133 - mod_facts should advertise TVFS support in the FEAT response.
- Bug 3128 - mod_sql_sqlite should use transactions for INSERT/UPDATE
  statements.
- Bug 3155 - Change the IdentLookups default to 'off'.  The RFC1413 IDENT
  lookup adds latency to the login process, so much so that it is a FAQ
  to configure "IdentLookups off".  In addition, the IDENT protocol is not
  secure; it can easily be spoofed using man-in-the-middle attacks.  Sites
  that require IDENT lookups must now explicitly configure "IdentLookups on".
- Bug 3156 - Allow resumed downloads when HiddenStore is in effect.
- Bug 3170 - RewriteMap unescape URL encoding broken by RewriteCondition
  backreference handling.
- Bug 2728 - Support for variables in ServerIdent directive.  The %L, %V,
  and %v variables are now supported for ServerIdent identity strings.
- Bug 3178 - mod_wrap2 does not handle IPv4-mapped IPv6 addresses.
- Bug 3179 - Restarted proftpd using mod_shaper will segfault.
- Bug 2720 - HideUser/HideGroup should be more flexible.
- Bug 3183 - Incorrect logging to wtmp.
- Bug 2608 - DNS should not be used for "server config" address discovery.
  A new -S command-line option has been added, for specifying the IP address
  to use for the host, rather than using DNS to resolve the host IP address
  from the hostname.
- Bug 3184 - When started in a chroot, proftpd fails to set transfer buffer
  size.
- Added mod_sftp, mod_sftp_pam, and mod_sftp_sql to the contrib modules.
- Bug 3185 - mod_wrap2 does not honour partial DNS name in access files.
  This appears to have been a regression caused by the fix for Bug#3178.
- Bug 3186 - Support use of <IfClass> directive with mod_ban.  The
  BanEngine directive can now appear in an <IfClass> section, in order
  to exclude connection classes from mod_ban's rules.
- Bug 3188 - prxs broken due to typo.
- Bug 3154 - Update bundled libtool version.  The bundled libltdl version
  has been updated to libtool 1.5.26.
- Bug 3111 - Support SSL/TLS server certificates stored in PKCS#12 files.
- Bug 3189 - Linker errors occur when using "make -j" parallel builds.
  ProFTPD now builds successfully using "make -jN", where N is the number
  of simultaneous build processes to use.
- Bug 3176 - Postgres chokes on standard charset names.
- Bug 3129 - Support configurable scoreboard scrubbing.  A new configuration
  directive, ScoreboardScrub, can be used to control whether the ScoreboardFile
  is scrubbed, and how often.
- Bug 3190 - MLSD/MLST do not honor <Limit> configurations.
- Bug 3196 - mod_quotatab does not honor last group in supplemental group list.
- Bug 3203 - Missing SQL backend modules can lead to null pointer segfault in
  mod_sql.
- Bug 3204 - Better support for installation and libs in prxs.  The prxs
  tool now handles the $Libraries$ build system hints in source files,
  and honors the DESTDIR environment variable.
- Bug 3167 - Support rewriting of home directories.
- Bug 2985 - Newest .spec file requires relatively current version of RPM.
  The provided .spec file now requires RPM 4.2 or later.
- Bug 3207 - Support for SSL/TLS session caching across processes/machines.
  The mod_tls module now supports a TLSSessionCache directive, and an API
  for modules to provide external SSL session caching mechanisms.
- Bug 3210 - Data transfers protected by SSL/TLS should reuse the control
  connection SSL session.  If the TLSRequired policy allows/requires SSL/TLS
  protections on data tranfers, then reuse of the SSL session from the
  control connection is now enforced.  Use "TLSOptions NoSessionReuseRequired"
  to relax this requirement.
- Bug 3215 - mod_wrap2_sql should support comma-delimited lists of clients.
  This is the same as Bug#3048, only it affects the mod_wrap2_sql module.
- Bug 2613 - Configure script should not append '/proftpd' to the
  --localstatedir.

1.3.2 - Released 5-Feb-2009
--------------------------------
- Bug 3169 - Multiple RewriteRules for the same RewriteCondition not processed
  properly.
- Bug 3171 - ExtendedLog should log full SITE command using %m.
- Bug 3173 - Encoding-dependent SQL injection vulnerability.

1.3.2rc4 - Released 23-Jan-2009
--------------------------------
- Bug 2045 - SQLShowInfo should not be displayed when query returns no data.
- Bug 2915 - mod_rewrite does not work well for SITE commands.
- Bug 1636 - GroupRatio does not check user's supplemental group membership.
- Bug 3137 - ProFTPD does not log filename %f for uploaded files.
- Bug 3142 - "Invalid number of arguments MFMT" due to spaces in path argument.
- Bug 3144 - mod_dynmasq returns same IP address, even though actual IP
  address has changed.
- Bug 3040 - Support for CreateHome parent directories owned by user.
- Added Russian translation.
- Bug 2020 - HideFiles sometimes fails.
- Bug 3146 - <Directory> paths using glob characters may not match as expected.
- Bug 3147 - Comma-delimited commands in <Limit> sections not handled properly.
- Bug 3149 - Bad handling of %p, %V, and %v variables in mod_sql.
- Bug 3150 - mod_facl erroneously assumes no permissions, rather than all
  permissions, in some cases.
- Bug 3159 - mod_rewrite build fails due to missing mode argument in open(2)
  call on some platforms.

1.3.2rc3 - Released 20-Nov-2008
--------------------------------

utils/ftpcount.c
/*
 * ProFTPD - FTP server daemon
 * Copyright (c) 1997, 1998 Public Flood Software
 * Copyright (c) 1999, 2000 MacGyver aka Habeeb J. Dihu <macgyver@tos.net>
 * Copyright (c) 2001-2008 The ProFTPD Project team
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307, USA.
 *
 * As a special exemption, Public Flood Software/MacGyver aka Habeeb J. Dihu
 * and other respective copyright holders give permission to link this program
 * with OpenSSL, and distribute the resulting executable, without including
 * the source code for OpenSSL in the source distribution.
 */

/* Shows a count of "who" is online via proftpd.  Uses the scoreboard file.
 *
 * $Id: ftpcount.c,v 1.17 2008-02-10 02:29:22 castaglia Exp $
 */

#include "utils.h"

#define MAX_CLASSES 100
struct scoreboard_class {
   char *score_class;
   unsigned long score_count;
};

static char *config_filename = PR_CONFIG_FILE_PATH;

static int check_scoreboard_file(void) {
  struct stat st;

  if (stat(util_get_scoreboard(), &st) < 0)
    return -1;

  return 0;
}

static struct option_help {
  char *long_opt, *short_opt, *desc;
} opts_help[] = {
  { "--config",	"-c",	"specify full path to proftpd configuration file" },
  { "--file",	"-f",	"specify full path to scoreboard file" },
  { "--help",	"-h",	NULL },
  { "--server",	"-S",	"show count only for specified ServerName" },
  { NULL }
};

#ifdef HAVE_GETOPT_LONG
static struct option opts[] = {
  { "config",  1, NULL, 'c' },
  { "file",    1, NULL, 'f' },
  { "help",    0, NULL, 'h' },
  { "server",  1, NULL, 'S' },
  { NULL,      0, NULL, 0   }
};
#endif /* HAVE_GETOPT_LONG */

static void show_usage(const char *progname, int exit_code) {
  struct option_help *h = NULL;

  printf("usage: %s [options]\n", progname);
  for (h = opts_help; h->long_opt; h++) {
#ifdef HAVE_GETOPT_LONG
    printf("  %s, %s\n", h->short_opt, h->long_opt);
#else /* HAVE_GETOPT_LONG */
    printf("  %s\n", h->short_opt);
#endif
    if (!h->desc)
      printf("    display %s usage\n", progname);
    else
      printf("    %s\n", h->desc);
  }

  exit(exit_code);
}

int main(int argc, char **argv) {
  pr_scoreboard_entry_t *score = NULL;
  pid_t oldpid = 0, mpid;
  time_t uptime = 0;
  unsigned int count = 0, total = 0;
  int c = 0, res = 0;
  char *server_name = NULL;
  struct scoreboard_class classes[MAX_CLASSES];
  char *cp, *progname = *argv;
  const char *cmdopts = "S:c:f:h";
  register unsigned int i;

  memset(classes, 0, MAX_CLASSES * sizeof(struct scoreboard_class));

  cp = strrchr(progname, '/');
  if (cp != NULL)
    progname = cp + 1;

  opterr = 0;
  while ((c =
#ifdef HAVE_GETOPT_LONG
	 getopt_long(argc, argv, cmdopts, opts, NULL)
#else /* HAVE_GETOPT_LONG */
	 getopt(argc, argv, cmdopts)
#endif /* HAVE_GETOPT_LONG */
	 ) != -1) {
    switch (c) {
      case 'h':
        show_usage(progname, 0);

      case 'f':
        util_set_scoreboard(optarg);
        break;

      case 'c':
        config_filename = strdup(optarg);
        break;

      case 'S':
        server_name = strdup(optarg);
        break;

      case '?':
        fprintf(stderr, "unknown option: %c\n", (char)optopt);
        show_usage(progname,1);
    }
  }

  /* First attempt to check the supplied/default scoreboard path.  If this is
   * incorrect, try the config file kludge.
   */
  if (check_scoreboard_file() < 0) {
    const char *file = util_scan_config(config_filename, "ScoreboardFile");
    if (file)
      util_set_scoreboard(file);

    if (check_scoreboard_file() < 0) {
      fprintf(stderr, "%s: %s\n", util_get_scoreboard(), strerror(errno));
      fprintf(stderr, "(Perhaps you need to specify the ScoreboardFile with -f, or change\n");
      fprintf(stderr," the compile-time default directory?)\n");
      exit(1);
    }
  }

  count = 0;
  if ((res = util_open_scoreboard(O_RDONLY)) < 0) {
    switch (res) {
      case UTIL_SCORE_ERR_BAD_MAGIC:
        fprintf(stderr, "error opening scoreboard: bad/corrupted file\n");
        return 1;

      case UTIL_SCORE_ERR_OLDER_VERSION:
        fprintf(stderr, "error opening scoreboard: bad version (too old)\n");
        return 1;

      case UTIL_SCORE_ERR_NEWER_VERSION:
        fprintf(stderr, "error opening scoreboard: bad version (too new)\n");
        return 1;

      default:
        fprintf(stderr, "error opening scoreboard: %s\n", strerror(errno));
        return 1;
    }
  }

  mpid = util_scoreboard_get_daemon_pid();
  uptime = util_scoreboard_get_daemon_uptime();

  errno = 0;
  while ((score = util_scoreboard_entry_read()) != NULL) {
    i = 0;

    if (errno)
      break;

    if (!count++ ||
        oldpid != mpid) {
      if (total)
        printf("   -  %d user%s\n\n", total, total > 1 ? "s" : "");

      if (!mpid)
        printf("inetd FTP connections:\n");
      else
        printf("Master proftpd process %u:\n", (unsigned int) mpid);

      if (server_name)
        printf("ProFTPD Server '%s'\n", server_name);

      oldpid = mpid;
      total = 0;
    }

    /* If a ServerName was given, skip unless the scoreboard entry matches. */
    if (server_name && strcmp(server_name, score->sce_server_label) != 0)
      continue;

    for (i = 0; i != MAX_CLASSES; i++) {
      if (classes[i].score_class == 0) {
	classes[i].score_class = strdup(score->sce_class);
	classes[i].score_count++;
        break;
      }

      if (strcasecmp(classes[i].score_class, score->sce_class) == 0) {
	classes[i].score_count++;
	break;
      }
    }

    total++;
  }
  util_close_scoreboard();

  /* Print out the total. */
  if (total) {
    for (i = 0; i != MAX_CLASSES; i++) {
      if (classes[i].score_class == 0)
         break;

      printf("Service class %-20s - %3lu %s\n", classes[i].score_class,
         classes[i].score_count, classes[i].score_count > 1 ? "users" : "user");
    }

  } else {
    if (!mpid)
      printf("inetd FTP connections:\n");
    else
      printf("Master proftpd process %u:\n", (unsigned int) mpid);

    printf("0 users\n");
  }

  return 0;
}
